packages/firestore/src/local/indexeddb_persistence.ts

```diff
@@ -1026,3 +1041,212 @@ function clientMetadataStore(
     DbClientMetadata.store
   );
 }
+
+/** Provides LRU functionality for IndexedDB persistence. */
+class IndexedDbLruDelegate implements ReferenceDelegate, LruDelegate {
+  private additionalReferences: ReferenceSet | null;
+
+  constructor(private readonly db: IndexedDbPersistence) {}
+
+  setInMemoryPins(inMemoryPins: ReferenceSet): void {
+    this.additionalReferences = inMemoryPins;
+  }
+
+  removeTarget(
+    txn: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
+    const updated = queryData.copy({
+      sequenceNumber: txn.currentSequenceNumber
+    });
+    return this.db.getQueryCache().updateQueryData(txn, updated);
+  }
+
+  addReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    return this.writeSentinelKey(txn, key);
+  }
+
+  removeReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    return this.writeSentinelKey(txn, key);
+  }
+
+  removeMutationReference(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    return this.writeSentinelKey(txn, key);
+  }
+
+  onLimboDocumentUpdated(
+    txn: PersistenceTransaction,
+    key: DocumentKey
+  ): PersistencePromise<void> {
+    return this.writeSentinelKey(txn, key);
+  }
+
+  forEachTarget(
+    txn: PersistenceTransaction,
+    f: (q: QueryData) => void
+  ): PersistencePromise<void> {
+    return this.db.getQueryCache().forEachTarget(txn, f);
+  }
+
+  forEachOrphanedDocumentSequenceNumber(
+    txn: PersistenceTransaction,
+    f: (sequenceNumber: ListenSequenceNumber) => void
+  ): PersistencePromise<void> {
+    return this.db
+      .getQueryCache()
+      .forEachOrphanedDocument(txn, (docKey, sequenceNumber) =>
+        f(sequenceNumber)
```

Comment:
Yes, there are no restrictions on duplicate sequence numbers, and this could in fact be common  if a bunch of documents were written in a single mutation. They would be ack'd in the same transaction, and so all get the same sequence number.